### PR TITLE
Feat rueidiscompat redisearch

### PR DESCRIPTION
**Pull request by unknowntpo**
Changes and cautions:
- docker-compose
  - add `compat-redisearch` service for testing `RediSearch` with `RESP2` 
- FT.CREATE:
  - search_command.json
	- Add options: `INDEXEMPTY`, `INDEXMISSING`, `WITHCOUNT`, `ADDSCORES`
	- Add field_type: `GEOSHAPE`
- FT.AGGREGATE:
  - search_command.json:
    - Add option: `WITHCOUNT`
- FT.SEARCH
  - go-redis doesn't implement `SUMMARIZE` option
  - Add option: `WITHCOUNT`

### CODE REVIEW

File: rueidiscompat/pipeline.go [hunk 1]
```diff
@@ -97,6 +97,8 @@ type Pipeline struct {
 	rets []Cmder
 }
 
+var pp = Pipeline{}
```
Comment (rueian) [inline rueidiscompat/pipeline.go:100]: What is this for?

File: rueidiscompat/adapter.go [hunk 2]
```diff
@@ -4835,6 +4835,615 @@ func (c *Compat) Watch(ctx context.Context, fn func(Tx) error, keys ...string) e
...
+				flatArgs := []interface{}{
+					"TYPE", sc.VectorArgs.FlatOptions.Type,
+					"DIM", sc.VectorArgs.FlatOptions.Dim,
+					"DISTANCE_METRIC", sc.VectorArgs.FlatOptions.DistanceMetric,
+				}
+				if sc.VectorArgs.FlatOptions.InitialCapacity > 0 {
+					flatArgs = append(flatArgs, "INITIAL_CAP", sc.VectorArgs.FlatOptions.InitialCapacity)
+				}
+				if sc.VectorArgs.FlatOptions.BlockSize > 0 {
+					flatArgs = append(flatArgs, "BLOCK_SIZE", sc.VectorArgs.FlatOptions.BlockSize)
```
Comment (rueian) [inline rueidiscompat/adapter.go:5132]: Lint Error: adapter.go:5132:6: this value of flatArgs is never used (SA4006)

File: rueidiscompat/pipeline.go (hunk 1 above)
Comment (unknowntpo) [reply rueidiscompat/pipeline.go]: removed.

File: rueidiscompat/adapter.go (hunk 2 above)
Comment (unknowntpo) [reply rueidiscompat/adapter.go:5171]: fixed.

File: message.go [hunk 3]
```diff
@@ -471,6 +471,16 @@ func (r RedisResult) ToAny() (v any, err error) {
...
 
+// IsMap delegates to RedisMessage.IsMap
+func (r RedisResult) IsMap() bool {
+	return r.val.IsMap()
+}
+
+// IsArray delegates to RedisMessage.IsArray
+func (r RedisResult) IsArray() bool {
+	return r.val.IsArray()
+}
```
Comment (rueian) [inline message.go:482]:
```suggestion
func (r RedisResult) isMap() bool {
	return r.val.IsMap()
}

// IsArray delegates to RedisMessage.IsArray
func (r RedisResult) isArray() bool {
	return r.val.IsArray()
}
```

Can we make these 2 methods private first? I am concerned about making these direct delegations to `r.val` as public APIs. I think we should at least ask users to check the `r.err` first.

File: rueidiscompat/util.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	t := time.NewTimer(dur)
+	defer t.Stop()
+
+	select {
+	case <-t.C:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
```
Comment (rueian) [inline rueidiscompat/util.go:47]: Can we remove unused functions in this file?

File: message.go [hunk 5]
```diff
@@ -456,7 +456,7 @@ func (r RedisResult) ToMap() (v map[string]RedisMessage, err error) {
 	if r.err != nil {
 		err = r.err
 	} else {
-		v, err = r.val.ToMap()
+		v, err = r.val.AsMap()
```
Comment (rueian) [inline message.go:459]: Why do we need this change? I think we already have an `AsMap` delegation.

Comment (unknowntpo) [reply message.go]: This is accidentally committed, I've reverted it.

File: message.go (hunk 3 above)
Comment (unknowntpo) [reply message.go:482]: But if we make `IsMap` private, functions in `command.go` can not access it,
I think we could check if any error exists, then delegate to `r.val.IsMap`.

```
func (r RedisResult) IsMap() bool {
	if r.err != nil {
		panic(fmt.Errorf("RedisResult has error: %w, please check the error before calling RedisResult.IsMap", r.err))
	}
	return r.val.IsMap()
}
```

File: rueidiscompat/util.go (hunk 4 above)
Comment (unknowntpo) [reply rueidiscompat/util.go]: Okay.

File: message.go (hunk 3 above)
Comment (unknowntpo) [reply message.go]: UPDATE: I'll use `res.ToMessage()` to get underlying message first, then do the `IsMap`, `IsArray` check.

**Review by rueian — APPROVED** [review]

**Pull request merged**